commons/com.b2international.index/src/com/b2international/index/query/AbstractExpressionBuilder.java

```diff
@@ -72,13 +78,90 @@ public B setMinimumNumberShouldMatch(int minShouldMatch) {
 	public Expression build() {
 		if (mustClauses.isEmpty() && mustNotClauses.isEmpty() && shouldClauses.isEmpty() && filterClauses.isEmpty()) {
 			return Expressions.matchAll();
-		} else if (mustClauses.isEmpty() && mustNotClauses.isEmpty() && shouldClauses.isEmpty() && filterClauses.size() == 1) {
+		} else if (isSingleFilter()) {
 			// shortcut to reduce number of nested Boolean clauses
 			return filterClauses.get(0);
```

Comment:
We could support `isSingleShould()` here as well, as you write in the final `else` block below (single "shoulds" _must_ match). 

Also if we want to be really precise, the extracted single filter clause should be wrapped in a non-scoring query wrapper to achieve the same effect as a boolean filter &ndash; which is not contributing to the final document score.

```suggestion
		} else if (isSingleFilter()) {
			// shortcut to reduce number of nested Boolean clauses
			return filterClauses.get(0);
			// return new NoScoreExpression(filterClauses.get(0)); <-- this class does not exist yet
		} else if (isSingleShould()) {
			return filterClauses.get(0);
```